django/db/models/deletion.py

```diff
@@ -117,6 +128,24 @@ def add_field_update(self, field, value, objs):
             model, {}).setdefault(
             (field, value), set()).update(objs)
 
+    def add_restricted_objects(self, field, objs):
+        if objs:
+            model = objs[0].__class__
+            self.restricted_objects.setdefault(model, {}).setdefault(field, set()).update(objs)
+
+    def clear_restricted_objects_from_set(self, model, objs):
+        self.restricted_objects[model] = {
+            field: items - objs for field, items in self.restricted_objects.get(model, {}).items()
+        }
+
+    def clear_restricted_objects_from_queryset(self, model, qs):
+        objs_to_clear = self.restricted_objects.get(model, {})
+        all_objs = set(qs.filter(pk__in=[o.pk for field in objs_to_clear for o in objs_to_clear[field]]))
+        self.restricted_objects[model] = {
+            field: items - all_objs
+            for field, items in self.restricted_objects.get(model, {}).items()
+        }
```

Comment:
IMO it's more readable without `get(model, {})` I would also collect `ID`'s in the first step, e.g.
```python
def clear_restricted_objects_from_queryset(self, model, qs):
    if model in self.restricted_objects:
        ids = [
            obj.pk
            for objs in self.restricted_objects[model].values() for obj in objs
        ]
        self.restricted_objects[model] = {
            field: items - set(qs.filter(pk__in=ids))
            for field, items in self.restricted_objects[model].items()
        }
```